### DISCUSSION TITLE
Questions about using the EMS-ESP API

### DISCUSSION THREAD

**Discussion by martinsamus** [emsesp/EMS-ESP32#1643] [Q&A] [answered]
Hi, I am using EMS-ESP together with Homey Pro. I set up MQTT to read out the data, MQTT didn't work for setting the data (I don't know why, but I've already given up on it, if there's someone who has dealt with it, they can write to me). Now I use API + HomeyScripts (using fetch command) to set and write data. But I have a question about boolean values.

If I have this EMS-ESP setup:
![image](https://github.com/emsesp/EMS-ESP32/assets/11224829/73ea39d6-a184-451c-a185-74318c22f0f0)

Thus, the response from the request http://ems-esp/api/thermostat/wwcharge shows "on" or "off" as a value:
```
{
     "name": "wwcharge",
     "fullname": "dhw charge",
     "circuit": "",
     "value": "on",
     "type": "boolean",
     "readable": true,
     "writeable": true,
     "visible": true
}
```

It's just formatting for display only, or for POST request as well? For example in request:
```
const body = {
   cmd:'wwcharge',
   value: 'off'
};
const res = await fetch('http://ems-esp/api/thermostat', {
         method: 'post',
         body: JSON.stringify(body),
         headers: {
           'Content-Type': 'application/json',
           'Authorization': 'Bearer ' + token
         },
     });
```
What value should I set "on"/"off" or true/false???

It is interesting that if I use the API to write the value "on" in WWCHARGE, it will start heating the water, but I cannot cancel it on the Junkers FW200 thermostat, or set it back to the value "off" by sending an "off" request. Still, if I look at the dashboard or read the value again using the API, the wwcharge value is still equal to the "on" value. This value is automatically switched off after one hour.
Changing the value to "off" does not work either using the web interface of the EMS-ESP application:
![image](https://github.com/emsesp/EMS-ESP32/assets/11224829/a9980bcb-ebf1-4d58-830a-43a83d033fca)

What should be the response to a POST request? I have this:
![image](https://github.com/emsesp/EMS-ESP32/assets/11224829/36bb990b-6dc5-4363-a951-230e612c0c96)

Thanks

**Comment by MichaelDvP** [comment, marked as answer]
> MQTT didn't work for setting the data (I don't know why, 

MQTT uses wildcard subscriptions, make sure your broker accept this. You have write to topic `device` or `device/entity`, not `device_data` as published. Wrting to topic `device` has to be json `{"cmd":"<cmd>", "data":<data>, "id":<n>}`, writing to `device/entity` can be a single value, "data" and "value" are equivalent, also "id" and "hc". Skipping hc defaults to first hc. 
e.g. `thermostat {"cmd":"seltemp", "data":20, "id":1}`, `thermostat/seltemp {"data":20}`, `thermostat/hc1/seltemp {"value":20}`, `thermostat/hc1/seltemp 20` are equivalent
> What value should I set "on"/"off" or true/false???

You can use always on/Off, true/false, 1/0 or the translated on/off in the selected language.
For enum values you can use the english or translated text or numbers, starting with 0.
>  it will start heating the water, but I cannot cancel it 

this is a function of the boiler, the command triggers a charge, the boiler charges until the set temperature is reached. You can lower the target temperature to stop charging. 
> What should be the response to a POST request?

on success only http 200 (ok), no body. On error http 400 (bad request) or 401 (unauthorized) and the error message in body.

**Reply by martinsamus** [reply]
Hi, what do I have set wrong? Setting the value works, but retrieving the status doesn't. Thank you

![Screenshot_20250114_023206_Samsung Internet.jpg](https://github.com/user-attachments/assets/749df157-37a9-4e14-a798-785c023c120d)


{
  "onoff": {
    "capability": "onoff",
    "stateTopic": "ems-esp/thermostat_data",
    "setTopic": "ems-esp/thermostat/dhw/charge",
    "valueTemplate": "$.dhw.charge",
    "outputTemplate": "value ? \"zap\" : \"vyp\"",
    "displayName": "Okamžitý ohrev TÚV"
  }
}

**Reply by proddy** [reply]
what do you mean? I tested in 3.7.2 and setting the entity language to DE and calling for example `http://ems-esp.local/api/boiler/burngas` shows the value for on/off is translated.

```json
{
  "name": "burngas",
  "fullname": "Gas",
  "circuit": "",
  "bool": false,
  "index": 0,
  "value": "aus",
  "type": "boolean",
  "readable": true,
  "writeable": false,
  "visible": true
}
```

**Reply by martinsamus** [reply]
Sorry, I didn't write it exactly. I use EMS-ESP integrated with Homey. I have created an OnOff button (MQTT Hub) and this is its definition:
```
{
   "onoff": {
      "capability": "onoff",
      "stateTopic": "ems-esp/thermostat_data",
      "setTopic": "ems-esp/thermostat/dhw/charge",
      "valueTemplate": "$.dhw.charge",
      "outputTemplate": "value ? "zap" : "vyp"",
      "displayName": "Okamžitý ohrev TÚV"
   }
}
```
As long as I have it set up like this, reading the current state (stateTopic) does not work. But if I press the button, the state on EMS-ESP changes to "zap" (SK language), but the button returns to the "not pressed/off/vyp" state.
Or another example:
Homey: the button is in the "not pressed/off/vyp" state. Using the application EMS-ESP, I change dhw.charge=zap, but the button defined on Homey is still in the "not pressed/off/vyp" state. This means that the changed value made in the EMS-ESP application is not read.
I am using stable version 3.7.1.
Thank you.

**Reply by proddy** [reply]
I did a quick test (and coded) it, and it seems to work. For example calling the API via an HTTP POST to the URL `/api` with a payload `"{"device":"boiler", "cmd":"circpump","value":"aus"}"` when the entity locale is set to DE

**Comment by martinsamus** [comment]
Thank you for your answer, I used MQTT as you wrote. But I'll probably have to find out if Homey Broker accepts wildcard subscriptions.
Thanks for the other answers as well.
